main.c

```diff
@@ -137,7 +142,58 @@ static void linux_boot(const struct boot_params boot_params)
     printf("  pref_address: 0x%llx\n", l_params->pref_address);
     printf("\n");
 
-    char *linux_header_end = (char *)(boot_params.kernel_addr + 0x201);
+    // We decided that the loader moves all kernels to their preferred address, even if they
+    // are relocatable.
+    uintptr_t actual_kernel_load_addr = boot_params.kernel_addr;
+
+    // Checks if a relocation of the kernel is required
+    if (boot_params.kernel_addr == actual_kernel_load_addr)
+    {
+        printf("The kernel is already loaded at its preferred address.\n");
+        printf("Booting kernel at 0x%lx\n", actual_kernel_load_addr);
+    } else {
+        printf("The kernel will be moved to its preferred load address.\n");
+        printf(
+            "Relocating from 0x%lx to %x%lx\n",
+            boot_params.kernel_addr,
+            actual_kernel_load_addr
+        );
+
+        uintptr_t kernel_begin_addr = actual_kernel_load_addr;
+        uintptr_t kernel_end_addr = kernel_begin_addr + l_params->init_size;
+
+        // check if the space is available in the memory map
+        for (int i = 0; i < lib_sysinfo.n_memranges; i++) {
+            die_on(kernel_begin_addr >= lib_sysinfo.memrange[i].base
+                       && kernel_end_addr < lib_sysinfo.memrange[i].base + lib_sysinfo.memrange[i].size
+                       && lib_sysinfo.memrange[i].type != CB_MEM_RAM,
+                   "the preferred load address is not in available RAM"
+            );
+        }
+
+        // check that the kernel will not clash with the initrd after relocation
+        uint64_t initrd_size = *(uint64_t *)boot_params.initrd_size_addr;
+        die_on(actual_kernel_load_addr >= boot_params.initrd_addr
+                   && actual_kernel_load_addr < boot_params.initrd_addr + initrd_size,
+               "Kernel load address collides with initrd\n"
+        );
+
+        // check that the kernel will not clash with the cmdline after relocation
+        uint64_t cmdline_size = strlen((char *)boot_params.cmdline_addr);
```

Comment:
use of `strlen` produces some bad feelings as there are safer alternatives.

Reply:
Below we explicitly do not look for the actual cmdline length. There isn't even a guarantee that this is properly null terminated. We could use a reasonable constant for the maximum size here.

> The number of kernel parameters is not limited, but the length of the complete command line (parameters including spaces etc.) is limited to a fixed number of characters. This limit depends on the architecture and is between 256 and 4096 characters. It is defined in the file ./include/asm/setup.h as COMMAND_LINE_SIZE.

```
arch/x86/include/asm/setup.h:#define COMMAND_LINE_SIZE 2048
```

Reply:
To increase actionability: Please use `strnlen` in combination with the useful max size that @snue suggested

Reply:
although... if it's just for the check, using `strnlen` is overkill.

Reply:
> although... if it's just for the check, using `strnlen` is overkill.

Yes. I recommend just checking against the constant and not even attempt to get the true command line length.